keyless/pepper/service/src/lib.rs

```diff
@@ -310,21 +315,61 @@ async fn update_account_recovery_db(input: &PepperInput) -> Result<(), Processin
                 aud: input.aud.clone(),
                 uid_key: input.uid_key.clone(),
                 uid_val: input.uid_val.clone(),
-                last_request_unix_ms: duration_since_epoch().as_millis() as u64,
+                first_request_unix_ms_minus_1q: None,
+                last_request_unix_ms: None,
+                num_requests: None,
             };
             let doc_id = entry.document_id();
-
-            // Actual DB operation using fluent API.
-            let op_result = db.fluent()
+            let now_unix_ms = duration_since_epoch().as_millis() as i64;
+
+            // The update transactions use the following strategy.
+            // 1. If not exists, create the document for the user identifier `(iss, aud, uid_key, uid_val)`.
+            //    but leave counter/time fields unspecified.
+            // 2. `num_requests += 1`, assuming the default value is 0.
+            // 3. `last_request_unix_ms = max(last_request_unix_ms, now)`, assuming the default value is 0.
+            // 4. `first_request_unix_ms = min(first_request_unix_ms, now)`, assuming the default value is +inf.
+            //
+            // This strategy is preferred because all the operations can be made server-side,
+            // which means the txn should require only 1 RTT,
+            // better than using read-compute-write pattern that requires 2 RTTs.
+            //
+            // This strategy does not work directly:
+            // in firestore, the default value of a number field is 0, and we do not know a way to customize it for op 4.
+            // The workaround here is apply an offset so 0 becomes a legitimate default value.
+            // So we work with `first_request_unix_ms_minus_1q` instead,
+            // which is defined as `first_request_unix_ms - 1_000_000_000_000_000`,
+            // where 1_000_000_000_000_000 milliseconds is roughly 31710 years.
```

Comment:
This small hack will affect the implementation of the recovery service: we gotta work with this new timestamp format carefully.

Are you sure you will not underflow?

Reply:
Range of i64 is [-2^63, 2^63-1], which includes -10^15.

firestore also uses i64
https://cloud.google.com/firestore/docs/concepts/data-types

Reply:
To get Unix time back, we only need to +10^15. Hopefully the recovery service extra work is minimized.